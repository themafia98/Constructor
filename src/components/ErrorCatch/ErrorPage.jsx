import React,{Fragment} from 'react';
import {withRouter} from 'react-router-dom';
import eventEmitter from '../../EventEmitter';

import './errorPage.scss';


class ErrorPage extends React.PureComponent {

    update = event => {
        this.props.history.push('/');
        eventEmitter.emit('EventErrorCatchUpdate');
    }

    render(){
        console.log('error');
        return (
        <div class = 'container'>
            <div class = 'flex-column'>
                <h1>Application crash</h1>
                <button onClick = {this.update}>Update</button>
            </div>
        </div>
        )
    }
}

export default withRouter(ErrorPage);