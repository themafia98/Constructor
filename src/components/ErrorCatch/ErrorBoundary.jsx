

import React from 'react';
import eventEmitter from '../../EventEmitter';
import ErrorPage from './ErrorPage';
import {Redirect} from 'react-router-dom';
class ErrorBoundary extends React.PureComponent {

      state = { 
        hasError: false,
      };

    static getDerivedStateFromError(error) {
        console.log(error);
      // Обновить состояние с тем, чтобы следующий рендер показал запасной UI.
      return { hasError: true};
    }

    updateApplication = event => {
      this.setState({hasError: false});
    };
  
    componentDidCatch(error, info) {
      // Можно также сохранить информацию об ошибке в соответствующую службу журнала ошибок
     console.error(error + "," + info);
    }
  
    render() {
      if (this.state.hasError) return <ErrorPage />
      return this.props.children;
    }

    componentDidMount = () => {
      eventEmitter.on('EventErrorCatchUpdate',this.updateApplication);
    }

    componentWillUnmount = () => {
      eventEmitter.off('EventErrorCatchUpdate',this.updateApplication);
    }
  }

  export default ErrorBoundary;