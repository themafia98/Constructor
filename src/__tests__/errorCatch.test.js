import React from 'react';
import renderer from 'react-test-renderer';

import ErrorBoundary from '../components/ErrorCatch/ErrorBoundary';
import ErrorPage from '../components/ErrorCatch/ErrorPage';

import Events from 'events';

let eventEmitter = new Events();

const ErrorBoundaryComponent = renderer.create(
    <ErrorBoundary>
            <ErrorPage />
    </ErrorBoundary>
);

test('Error catch test', () => {

    let snapshotErrorBoundaryComponent = ErrorBoundaryComponent.toJSON();
    expect(snapshotErrorBoundaryComponent).toMatchSnapshot();
    // try catch errors
    const updateApp = ErrorBoundaryComponent.root.find(item => item.props.className === 'ErrorPage__button_updateApp');
    updateApp.props.onClick({stopPropagation: () => null});

    snapshotErrorBoundaryComponent = ErrorBoundaryComponent.toJSON();
    expect(snapshotErrorBoundaryComponent).toMatchSnapshot();
});