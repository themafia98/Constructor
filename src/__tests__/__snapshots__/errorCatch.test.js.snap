// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error catch test 1`] = `
<div
  class="container"
>
  <div
    class="flex-column ErrorPage"
  >
    <h1>
      Application crash
    </h1>
    <button
      className="ErrorPage__button_updateApp"
      onClick={[Function]}
    >
      Update
    </button>
  </div>
</div>
`;

exports[`Error catch test 2`] = `
<div
  class="container"
>
  <div
    class="flex-column ErrorPage"
  >
    <h1>
      Application crash
    </h1>
    <button
      className="ErrorPage__button_updateApp"
      onClick={[Function]}
    >
      Update
    </button>
  </div>
</div>
`;
